scalableminds/webknossos
# fix massive saving ddos

## What happened?
The frontend occasionally sends several save request within a one second interval. These peaks mess up stuff on the server. 😡 How is this happening?

1. The client stores all updates/actions in a save queue.
2. The `save_saga` empties this queue every 30s (or when the user click the save now button)
3. The actual response is fired from within a `while` loop which will run until the save queue has been emptied.
4. Simultaneously more update/actions are added to the queue. Especially tracing in flight mode fills up the queue really quickly. This leads to situations, where after the save request has been send the 
queue already has new items.
5. Given that the request is fired from within a `while` loop which is trying to empty the queue, the loop will send continuously send save requests as quickly as possible to empty the queue. ⚡️ Many many requests are send DDOS style.

## How is it fixed?
~~I remove the while loop. Now a save request will definitely only run once and save as much updates/actions as possible but no more. If the queue fills up again in the meantime, so be it. The queue will be emptied/saved the in the next 30s again.~~

I kept the while loop but I break after one successful save request. Why keep the loop? In case the request fails it will be tried again.



Mailable description of changes (needs to be understandable by webknossos mailing list people):
- fix the massive amounts of saving requests that happen sometimes

Steps to test:
- Open Network Tab
- Trace (better yet use flight mode) for some time
- Notice that save request are only send every 30s

------
